de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java

```diff
@@ -0,0 +1,164 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * Class used to capture and re-apply which editors are currently selected by
+ * the user.
+ */
+//TODO consider duplicated open editors during screen splitting
+public class SelectedEditorState {
+    private static final Logger log = Logger
+        .getLogger(SelectedEditorState.class);
+
+    private final List<VirtualFile> selectedEditors;
+
+    private boolean hasCapturedState;
+
+    @Inject
+    private ProjectAPI projectAPI;
+
+    @Inject
+    private EditorManager editorManager;
+
+    @Inject
+    private Project project;
+
+    public SelectedEditorState() {
+        this.selectedEditors = new ArrayList<>();
+        this.hasCapturedState = false;
+
+        SarosPluginContext.initComponent(this);
+    }
+
+    /**
+     * Captures the local selected editor state.
+     */
+    public void captureState() {
+        if (hasCapturedState) {
+            log.warn("Overwriting existing captured selected editor state.");
+
+            selectedEditors.clear();
+        }
+
+        selectedEditors.addAll(Arrays.asList(projectAPI.getSelectedFiles()));
+
+        hasCapturedState = true;
+    }
+
+    /**
+     * Applies the captured selected editor state to the local IDE.
+     */
+    public void applyCapturedState() {
+        if (!hasCapturedState) {
+            log.warn(
+                "Trying to applying state before capturing a local state to"
+                    + "re-apply.");
+
+            return;
+        }
+
+        ListIterator<VirtualFile> iterator = selectedEditors
+            .listIterator(selectedEditors.size());
+
+        try {
+            editorManager.getFileListener().unsubscribe();
+
+            while (iterator.hasPrevious()) {
+                projectAPI.openEditor(iterator.previous(), true);
+            }
+
+        } finally {
+            editorManager.getFileListener().subscribe(project);
+        }
+    }
+
+    /**
+     * Replaces the given old file with the given new file in the captured
+     * selected editor state.
+     *
+     * @param oldFile the old file to remove from the captured selected editor
+     *                state
+     * @param newFile the new file to add to the captured selected editor state
+     *                in the position of the old file
+     */
+    public void replaceSelectedFile(
+        @NotNull
+            VirtualFile oldFile,
+        @NotNull
+            VirtualFile newFile) {
+
+        if (!hasCapturedState) {
+            log.warn("Trying to replace file in state before capturing a local "
+                + "state. old file: " + oldFile + ", new file: " + newFile);
+
+            return;
+        }
+
+        int index = selectedEditors.indexOf(oldFile);
+
+        if (index != -1) {
+            selectedEditors.set(index, newFile);
+
+        } else {
+            log.debug("Could not replace " + oldFile + " with " + newFile
```

Comment:
I changed this log level to debug as there is currently no good logic to check whether a given file is selected (without involving a lot of IntelliJ logic). This means that this method will always be called, independent of whether the file to replace was actually selected to begin with.

I decided to do it this way around as the initial check whether the file is selected would probably take up more computing time than unnecessarily calling this method.

The only somewhat ugly remain is that this message will be logged in such cases, making it harder to determine when this case is actually a problem (caused by an inconsistency of some kind). But this is acceptable in my opinion.